### ISSUE TITLE
save_field hook throws a fatal TypeError

### ISSUE THREAD

**Issue by noplanman** [CMB2/CMB2#1409]
## Describe the bug

When trying to save a text field that has no value, I got this error:
```
PHP message: PHP Fatal error:  Uncaught TypeError: Argument 2 passed to {closure}() must be of the type bool, array given
```

I'm using the `save_field` action, and the error occurs in that closure:
```php
add_action( 'cmb2_save_field', static function ( string $field_id, bool $updated ) {
    ...
}
```

## Steps to reproduce
Not entirely sure, as this is only happening on one of my projects.

What I have is a list of fields, and having one of the text ones empty produces the error.

When I add a dummy value, it saves correctly. Removing it again, I get the fatal error.

## Possible Solution
`$updated` should be a boolean, but `CMB2_Field::remove_data()` ([here](https://github.com/CMB2/CMB2/blob/5efff8833ad1c3888feb7e0df421554e2a06f957/includes/CMB2_Field.php#L572)) doesn't necessarily return a boolean, so the `do_action` calls further down fail, due to passing `array` instead of `boolean`: 
- https://github.com/CMB2/CMB2/blob/5efff8833ad1c3888feb7e0df421554e2a06f957/includes/CMB2_Field.php#L593
- https://github.com/CMB2/CMB2/blob/5efff8833ad1c3888feb7e0df421554e2a06f957/includes/CMB2_Field.php#L607

### Possible Solution's Risk Level
For the `do_action` calls, cast to `boolean`. Since the returned value of `remove_data` is an array (which contains all the saved values of the fields), this would cast to true, but that may not necessarily be true.

## Your Environment
WP 5.6.1
CMB2 version 2.8.0

**Comment by tw2113** [comment]
This is from your code, correct?

```
add_action( 'cmb2_save_field', static function ( string $field_id, bool $updated ) {
    ...
}
```
I'm looking at the `save_field()` method in CMB2 which is where the `cmb2_save_field` hook is in, and I'm seeing possible return values of bool, int, and array. Thus it feels much less of a CMB2 "bug" of sorts and more a parameter type cast detail with the callback above.

**Comment by noplanman** [comment]
@tw2113 Correct, that is my code.

Going by the Docblock of the action, the parameters are typed, so one would expect the passed variables to match that typing, no?

Else the Docblock would need to be updated to include the other types.

I accept that I may need to be more liberal with the typing, if this is the correct behaviour 😊

**Comment by tw2113** [comment]
Yeah, the docblock is definitely not accurate in this case. For example [lines 555-565](https://github.com/CMB2/CMB2/blob/develop/includes/CMB2_Field.php#L555-L565) shows iterating on an integer that then gets assigned to the `$updated` variable as well.

**Comment by noplanman** [comment]
Right. Just seems strange to be getting an array of all the field values, when no value is changed in a given text field 🤔 
Because even then, casting to boolean (which will become `true`) isn't correct.
And assuming `false` when an array gets passed also seems wrong 🤷

**Comment by tw2113** [comment]
Seeing if i can maybe circle back to this later in the week to touch up the docs for the hook, and maybe review other hook documentation at the same time. The more I poke through the code, the more variant in return types I'm seeing, and less ease in type casting at the end, without risking breaking things for others.
